service/src/main/kotlin/fi/espoo/evaka/invoicing/service/InvoiceGenerator.kt

```diff
@@ -69,21 +70,60 @@ class InvoiceGenerator(
             tracer.withSpan("applyCorrections") {
                 applyUnappliedCorrections(tx, month, invoices, invoiceCalculationData.areaIds)
             }
+
         tx.deleteDraftInvoices(month, InvoiceStatus.DRAFT)
         tx.insertDraftInvoices(
             status = InvoiceStatus.DRAFT,
             invoices = invoicesWithCorrections,
-            relatedFeeDecisions =
-                invoicesWithCorrections.associate { invoice ->
-                    invoice.headOfFamily to
-                        invoiceCalculationData.decisions
-                            .getOrDefault(invoice.headOfFamily, emptyList())
-                            .map { it.id }
-                },
+            relatedFeeDecisions = invoiceCalculationData.decisionIds,
         )
     }
 
     fun generateAllReplacementDraftInvoices(dbc: Database.Connection, today: LocalDate) {
+        forReplaceableMonths(dbc, today) { tx, month ->
+            val invoiceCalculationData =
+                tracer.withSpan("calculateInvoiceData") {
+                    calculateInvoiceData(tx, month, excludeAlreadyInvoiced = false)
+                }
+            val drafts = createReplacementDraftInvoices(tx, month, invoiceCalculationData)
+
+            tx.deleteDraftInvoices(month, InvoiceStatus.REPLACEMENT_DRAFT)
+            tx.insertDraftInvoices(
+                status = InvoiceStatus.REPLACEMENT_DRAFT,
+                invoices = drafts,
+                relatedFeeDecisions = invoiceCalculationData.decisionIds,
+            )
+        }
+    }
+
+    fun generateReplacementDraftInvoicesForHeadOfFamily(
+        dbc: Database.Connection,
+        today: LocalDate,
+        headOfFamilyId: PersonId,
+    ) =
+        forReplaceableMonths(dbc, today) { tx, month ->
+            val invoiceCalculationData =
+                tracer.withSpan("calculateInvoiceData") {
+                    calculateInvoiceData(tx, month, headOfFamilyId)
+                }
+            val draft =
+                createReplacementDraftInvoices(tx, month, invoiceCalculationData).firstOrNull()
```

Comment:
started thinking that a generic extension function `.exactlyOneOrNull()` for iterables might be useful to not hide bugs